[DAG] Fold (umin (sub a b) a) -> (usubo a b); (select usubo.1 a usubo.0)

ckoparkar opened the pull request:
Fixes #161036.

ckoparkar committed:
[DAG] Fold (umin (sub a b) a) -> (usubo a b); (select usubo.1 a usubo.0)

ckoparkar committed:
Remove smin pattern, it might not be correct

ckoparkar committed:
Merge branch 'main' into ckoparkar/161036

ckoparkar committed:
Fix check for retq

ckoparkar commented:
/cc @RKSimon

ckoparkar committed:
Merge branch 'main' into ckoparkar/161036

RKSimon reviewed (changes requested):
please add aarch64 test coverage

ckoparkar commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -6199,6 +6199,28 @@ SDValue DAGCombiner::visitIMINMAX(SDNode *N) {
+  if (N0.getOpcode() == ISD::SUB) {
+    SDValue A, B, C;
+    if (sd_match(N, m_UMin(m_Sub(m_Value(A), m_Value(B)), m_Value(C)))) {
+      EVT AVT = A.getValueType();
```
That's true actually, can reuse VT.

ckoparkar committed:
Address review comments

ckoparkar commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -6199,6 +6199,28 @@ SDValue DAGCombiner::visitIMINMAX(SDNode *N) {
+    if (sd_match(N, m_UMin(m_Sub(m_Value(A), m_Value(B)), m_Value(C)))) {
+      EVT AVT = A.getValueType();
+      if (A == C && TLI.isOperationLegalOrCustom(ISD::USUBO, AVT)) {
+        SDVTList VTs = DAG.getVTList(AVT, MVT::i1);
```
It'll always be MVT::i1 unless umin is applied on a vector? I'll add a `!VT.isVector()` guard.

ckoparkar committed:
Don't use MVT::i1

RKSimon reviewed (changes requested):
Please can you update against trunk latest and investigate the CI failure

ckoparkar committed:
Fix pattern match

ckoparkar committed:
Add some more tests

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -6199,6 +6199,19 @@ SDValue DAGCombiner::visitIMINMAX(SDNode *N) {
+  {
+    SDValue A, B;
+    if (sd_match(N0, m_Sub(m_Value(A), m_Value(B))) &&
+        sd_match(N1, m_Specific(A)) &&
```
Remove A entirely if its supposed to match N1?
```
if (sd_match(N0, m_Sub(m_Specific(N1), m_Value(B))) &&
```

ckoparkar commented:
1. Test failure

<details>

Previously the code was:

```
if (N0.getOpcode() == ISD::SUB) {
  sd_match(N, m_UMin(m_Sub(m_Value(A), m_Value(B)), m_Deferred(A)))
  ...
```
Once I removed the `if (N0.getOpcode() == ISD::SUB)`, this pattern match grabbed this:

```
  SelectionDAG has 17 nodes:
  t2: i32,ch = CopyFromReg t0, Register:i32 %0
                t8: i32 = sub t2, Constant:i32<4>
              t9: i32 = umin t2, t8


N: t9: i32 = umin t2, t8
A: t2: i32,ch = CopyFromReg t0, Register:i32 %0
B: t7: i32 = Constant<4>
```

`t8: i32 = sub..` is the second argument of `umin` so I'm kind of confused as to why this pattern match succeeded. I'm now pattern matching on individual arguments of `umin`.

</details>

2. More tests

<details>

I've added some tests with vector types, incorrect patterns, etc. @arsenm also mentioned a test for multiple use.

I didn't guard against multiple uses here since I thought the results of `sub` and `umin` are still available after the fold. Am I misunderstanding something here and should I be guarding the pattern with `m_OneUse`?

</details>

ckoparkar committed:
Simplify pattern match

ckoparkar committed:
Merge branch 'main' into ckoparkar/161036

* main: (3124 commits)
  [X86] narrowBitOpRMW - add handling for single bit insertion patterns (#165742)
  [gn build] Port 5322fb626820
  [libc++] Simplify the implementation of destroy_at a bit (#165392)
  [MLIR][NVVM] Update mbarrier.init/inval Ops to use AnyTypeOf[] (#165558)
  [X86] Remove AMX-TRANSPOSE (#165556)
  [CIR] Fix multiple returns in switch statements (#164468)
  [lld][test] Fix file cleanup in aarch64-build-attributes.s (#164396)
  [X86] combineTruncate - trunc(srl(load(p),amt)) -> load(p+amt/8) - ensure there isn't an interdependency between the load and amt (#165850)
  [llvm][docs] Remove guidance on adding release:reviewed label (#164395)
  [libc++] Update our documentation on the supported compilers (#165684)
  [AMDGPU][GlobalISel] Add register bank legalization for G_FADD (#163407)
  [LLVM][ConstantFolding] Extend constantFoldVectorReduce to include scalable vectors. (#165437)
  [SDAG] Set InBounds when when computing offsets into memory objects (#165425)
  [llvm][DebugInfo][ObjC] Make sure we link backing ivars to their DW_TAG_APPLE_property (#165409)
  [NFCI] Address post-merge review of #162503 (#165582)
  [clang][tools][scan-view] Remove Python2 compatibility code in ScanView.py (#163747)
  [lldb][TypeSystem] Remove count parameter from TypeSystem::IsFloatingPointType (#165707)
  [llvm][tools][opt-viewer] Put back missing function
  [llvm][tools][opt-viewer] Remove Python2 compatability code in optrecord.py (#163744)
  [clang][utils] Make CmpDriver Python3 compatible (#163740)
  ...

ckoparkar commented on llvm/test/CodeGen/AArch64/underflow-compare-fold.ll:
```diff
@@ -0,0 +1,14 @@
+  %sub = sub i64 %a, %b
+  %cond = tail call i64 @llvm.umin.i64(i64 %sub, i64 %a)
+  ret i64 %cond
+}
```
I've added some tests with vector types, incorrect patterns, etc.

I didn't guard against multiple uses here since I thought the results of sub and umin are still available after the fold. Am I misunderstanding something here and should I be guarding the pattern with m_OneUse?

arsenm replied:
The point isn't whether it's correct or not, but whether it's worthwhile. If you aren't eliminating the instruction, is it worth doing the fold (I suppose maybe if the min will be expanded)

ckoparkar committed:
Merge branch 'main' into ckoparkar/161036

* main:
  [SPIRV] Fix vector bitcast check in LegalizePointerCast (#164997)
  [lldb][docs] Add troubleshooting section to scripting introduction
  [Sema] Fix parameter index checks on explicit object member functions (#165586)
  To fix polymorphic pointer assignment in FORALL when LHS is unlimited polymorphic and RHS is intrinsic type target (#164999)
  [CostModel][AArch64] Model cost of extract.last.active intrinsic (clastb) (#165739)
  [MemProf] Select largest of matching contexts from profile (#165338)
  [lldb][TypeSystem] Better support for _BitInt types (#165689)
  [NVPTX] Move TMA G2S lowering to Tablegen (#165710)
  [MLIR][NVVM] Extend NVVM mma ops to support fp64 (#165380)
  [UTC] Support to test annotated IR (#165419)

ckoparkar committed:
Merge branch 'main' into ckoparkar/161036

arsenm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -6216,6 +6216,18 @@ SDValue DAGCombiner::visitIMINMAX(SDNode *N) {
                                         SDLoc(N), VT, N0, N1))
     return SD;
 
+  // (umin (sub a, b) a) -> (usubo a, b); (select usubo.1, a, usubo.0)
```
```suggestion
  // (umin (sub a, b), a) -> (usubo a, b); (select usubo.1, a, usubo.0)
```

arsenm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -6216,6 +6216,18 @@ SDValue DAGCombiner::visitIMINMAX(SDNode *N) {
+    if (sd_match(N0, m_Sub(m_Specific(N1), m_Value(B))) &&
+        TLI.isOperationLegalOrCustom(ISD::USUBO, VT)) {
+      EVT SETCCT = getSetCCResultType(VT);
+      SDVTList VTs = DAG.getVTList(VT, SETCCT);
```
```suggestion
      SDVTList VTs = DAG.getVTList(VT, getSetCCResultType(VT));
```

arsenm reviewed (commented):
This probably isn't profitable if the min is legal. e.g., for AMDGPU I expect this to be profitable for i64 but not i32

ckoparkar committed:
Edits

ckoparkar committed:
Merge branch 'main' into ckoparkar/161036

* main: (1028 commits)
  [clang][DebugInfo] Attach `DISubprogram` to additional call variants (#166202)
  [C2y] Claim nonconformance to WG14 N3348 (#166966)
  [X86] 2012-01-10-UndefExceptionEdge.ll - regenerate test checks (#167307)
  Remove unused standard headers: <string>, <optional>, <numeric>, <tuple> (#167232)
  [DebugInfo] Add Verifier check for incorrectly-scoped retainedNodes (#166855)
  [VPlan] Don't apply predication discount to non-originally-predicated blocks (#160449)
  [libc++] Avoid overloaded `operator,` for (`T`, `Iter`) cases (#161049)
  [tools][llc] Make save-stats.ll test target independent (#167238)
  [AArch64] Fallback to PRFUM for PRFM with negative or unaligned offset (#166756)
  [X86] ldexp-avx512.ll - add v8f16/v16f16/v32f16 test coverage for #165694 (#167294)
  [DropAssumes] Drop dereferenceable assumptions after vectorization. (#166947)
  [VPlan] Simplify branch-cond with getVectorTripCount (#155604)
  Remove unused <algorithm> inclusion (#166942)
  [AArch64] Combine subtract with borrow to SBC. (#165271)
  [AArch64][SVE] Avoid redundant extend of unsigned i8/i16 extracts. (#165863)
  [SPIRV] Fix failing assertion in SPIRVAsmPrinter (#166909)
  [libc++] Merge insert/emplace(const_iterator, Args...) implementations (#166470)
  [libc++] Replace __libcpp_is_final with a variable template (#167137)
  [gn build] Port 152bda726958
  [libc++] Replace the last uses of __tuple_types with __type_list (#167214)
  ...

ckoparkar committed:
Adjust patterns for AMDGPU

ckoparkar committed:
Adjust patterns in AMDGPU/llvm.set.rounding.ll

ckoparkar committed:
Apply arsenm's suggestion

ckoparkar commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -6219,6 +6219,25 @@ SDValue DAGCombiner::visitIMINMAX(SDNode *N) {
                                         SDLoc(N), VT, N0, N1))
     return SD;
 
+  if (TLI.isOperationLegalOrCustom(ISD::USUBO, VT) &&
```
Um wouldn't we want to check if `usubo` is legal?

arsenm replied:
Mostly I'm wondering how to get this to fire in the 64-bit amdgpu case. usubo isn't legal, but expanding the 64-bit usubo is as good as expanding the min

ckoparkar commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -6219,6 +6219,24 @@ SDValue DAGCombiner::visitIMINMAX(SDNode *N) {
                                         SDLoc(N), VT, N0, N1))
     return SD;
 
+  if (TLI.isOperationLegalOrCustom(ISD::USUBO, VT)) {
```
Thanks I've applied your suggestion :-) The AMDGPU code is as expected now, the fold doesn't apply essentially.

ckoparkar committed:
Merge branch 'main' into ckoparkar/161036

* main: (63 commits)
  [libc++] Inline vector::__append into resize (#162086)
  [Flang][OpenMP] Move char box bounds generation for Maps to DirectiveCommons.h (#165918)
  RuntimeLibcalls: Add entries for vector sincospi functions (#166981)
  [X86] _mm_addsub_pd is not valid for constexpr (#167363)
  [CIR] Re-land: Recognize constant aggregate initialization of auto vars (#167033)
  [gn build] Port d2521f14a73d
  [gn build] Port caed0893465a
  [gn build] Port 315d705b7608
  [gn build] Port 2345b7d98f75
  [PowerPC] convert memmove to milicode call  .___memmove64[PR]  in 64-bit mode (#167334)
  [HLSL] Add internal linkage attribute to resources (#166844)
  AMDGPU: Update test after e95f6fa12
  [bazel] Port #166980: TLI/VectorLibrary refactor (#167354)
  [libc++] Split macros related to hardening into their own header (#167069)
  [libc++][NFC] Remove unused imports from generate_feature_test_macro_components.py (#159591)
  [CIR][NFC] Add test for Complex imag with GUN extension (#167215)
  [BOLT][AArch64] Add more heuristics on epilogue determination (#167077)
  RegisterCoalescer: Enable terminal rule by default for AMDGPU (#161621)
  Revert "[clang] Refactor option-related code from clangDriver into new clangOptions library" (#167348)
  [libc++][docs] Update to refer to P3355R2 (#167267)
  ...

RKSimon reviewed (approved):
LGTM

mstorsjo commented:
This change broke a significant number of cases in ffmpeg, across multiple architectures.

One way of reproducing is this:
```console
$ git clone https://github.com/ffmpeg/ffmpeg
$ mkdir ffmpeg-build
$ cd ffmpeg-build
$ ../ffmpeg/configure --cc=clang
$ make -j$(nproc)
$ make fate-vsynth1-flv
```
This produces a binary that crashes for this testcase.

On x86_64, the object file where the breakage is, is in `libavcodec/x86/videodsp_init.o`. On aarch64, the fault appears in `libavcodec/videodsp.o`.

Compilation of that particular object file (the one on x86) can be done standalone with https://martin.st/temp/videodsp_init-preproc.c - with `clang -target x86_64-linux-gnu -c -O2 videodsp_init-preproc.c`.

If comparing the difference in generated code caused by this patch, I get this:
```diff
--- out-good.s  2025-11-13 11:30:40.037589941 +0200
+++ out-bad.s   2025-11-13 11:30:33.665583286 +0200
@@ -132,15 +132,13 @@
        addq    %r10, %r11
        movq    %rax, %rbx
 .LBB1_12:                               # %if.end29.i
+       xorl    %eax, %eax
        subq    %r8, %r9
        negq    %r8
-       xorl    %eax, %eax
-       testq   %r8, %r8
-       cmovleq %rax, %r8
+       cmovbq  %rax, %r8
        movq    %rbx, %r13
        negq    %r13
-       testq   %r13, %r13
-       cmovleq %rax, %r13
+       cmovbq  %rax, %r13
        cmpq    %r9, %r14
        cmovlq  %r14, %r9
        subq    %rbx, %rbp
@@ -374,15 +372,13 @@
        addq    %r10, %r11
        movq    %rax, %rbx
 .LBB2_12:                               # %if.end29.i
+       xorl    %eax, %eax
        subq    %r8, %r9
        negq    %r8
-       xorl    %eax, %eax
-       testq   %r8, %r8
-       cmovleq %rax, %r8
+       cmovbq  %rax, %r8
        movq    %rbx, %r13
        negq    %r13
-       testq   %r13, %r13
-       cmovleq %rax, %r13
+       cmovbq  %rax, %r13
        cmpq    %r9, %r14
        cmovlq  %r14, %r9
        subq    %rbx, %rbp
```


I would suggest reverting this commit for now.
